gossip/services/tcp_receiver.cc

```diff
@@ -19,54 +22,101 @@ static logger_id_t logger_id;
+           remote_host_.c_str(), remote_port_);
 }
 
 void TcpConnection::start() {
-  auto host = socket_.remote_endpoint().address().to_string().c_str();
-  auto port = socket_.remote_endpoint().port();
+  remote_host_ = socket_.remote_endpoint().address().to_string();
 
-  rw_lock_handle_rdlock(&service_->state->node->neighbors_lock);
+  // Reading listening port from node
+
+  std::array<char, PORT_SIZE> port_bytes;
+  boost::system::error_code error;
+  size_t len = socket_.read_some(boost::asio::buffer(port_bytes), error);
+  if (len != PORT_SIZE) {
+    log_warning(logger_id, "Received invalid port from node tcp://%s:%d\n",
+                remote_host_.c_str(), remote_port_);
+    return;
+  }
+  remote_port_ = std::stoi(std::string(&port_bytes[0], len));
+
+  // Looking for matching neighbor
+
+  rw_lock_handle_wrlock(&service_->state->node->neighbors_lock);
 
   neighbor_t* neighbor = neighbors_find_by_endpoint_values(
-      service_->state->node->neighbors, host, port, PROTOCOL_TCP);
+      service_->state->node->neighbors, remote_host_.c_str(), remote_port_,
+      PROTOCOL_TCP);
 
   if (neighbor == NULL) {
     log_info(logger_id,
-             "Connection denied with non-tethered neighbor tcp://%s:%d\n", host,
-             port);
+             "Connection denied with non-tethered neighbor tcp://%s:%d\n",
+             remote_host_.c_str(), remote_port_);
     rw_lock_handle_unlock(&service_->state->node->neighbors_lock);
     return;
   }
 
   log_info(logger_id,
-           "Connection accepted with tethered neighbor tcp://%s:%d\n", host,
-           port);
+           "Connection accepted with tethered neighbor tcp://%s:%d\n",
+           remote_host_.c_str(), remote_port_);
+
+  // Opening connection with neighbor
+
+  boost::asio::ip::tcp::endpoint neighbor_endpoint(
+      boost::asio::ip::address::from_string(remote_host_), remote_port_);
+  auto neighbor_socket =
+      new boost::asio::ip::tcp::socket(socket_.get_executor().context());
+  neighbor_socket->connect(neighbor_endpoint);
+
+  // Sending listening port to neighbor
+
+  std::stringstream stream;
+  stream << std::setfill('0') << std::setw(PORT_SIZE) << remote_port_;
+  boost::asio::write(*neighbor_socket, boost::asio::buffer(stream.str()),
+                     error);
+
+  neighbor->endpoint.opaque_inetaddr = neighbor_socket;
+
   rw_lock_handle_unlock(&service_->state->node->neighbors_lock);
 
-  receive();
-}
+  iota_packet_t packet;
+
+  for (;;) {
+    std::array<char, PACKET_SIZE + CRC_SIZE> tcp_packet;
+
+    boost::asio::read(socket_, boost::asio::buffer(tcp_packet), error);
+
+    if (error) {
+      log_warning(logger_id,
+                  "Reading from tethered node tcp://%s:%d failed: %s\n",
+                  remote_host_.c_str(), remote_port_, error.message().c_str());
+      break;
+    }
+
+    // Computing CRC
+
+    boost::crc_32_type result;
+    result.process_bytes(&tcp_packet[0], PACKET_SIZE);
+    stream.clear();
```

Comment:
no need for this.
just compare if provided crc matches calculated
also don't use a stream?